# Introdução às Strings em Python
Strings são sequências de caracteres e são usadas para representar texto em Python. Podemos criar strings usando aspas simples ou duplas. Vamos explorar como criar e acessar strings em Python.

In [ ]:
# Criando Strings
string1 = 'Olá, Mundo!'
string2 = "Python é divertido!"

# Acessando Caracteres
primeiro_caractere = string1[0]
ultimo_caractere = string2[-1]

string1, string2, primeiro_caractere, ultimo_caractere

('Olá, Mundo!', 'Python é divertido!', 'O', '!')

## Operações Básicas com Strings
Podemos realizar várias operações com strings, como concatenar duas strings, repetir uma string e verificar se uma substring está presente em outra string. Vamos ver alguns exemplos dessas operações.

In [ ]:
# Concatenação de Strings
concatenada = string1 + ' ' + string2

# Repetição de String
repetida = string1 * 3

# Verificação de Substring
substring_presente = 'Mundo' in string1

concatenada, repetida, substring_presente

('Olá, Mundo! Python é divertido!', 'Olá, Mundo!Olá, Mundo!Olá, Mundo!', True)

## Métodos de Strings
As strings em Python vêm com vários métodos úteis que nos permitem realizar operações comuns de maneira fácil e eficiente. Vamos explorar alguns desses métodos.

| Método                     | Exemplo com Saída                                              | Descrição                                                                                          |
|----------------------------|----------------------------------------------------------------|-----------------------------------------------------------------------------------------------------|
| `str.capitalize()`         | `("a vida é bela.").capitalize()` => "A vida é bela."          | Converte a primeira letra da string em maiúscula.                                                   |
| `str.lower()`              | `("A Vida é bela.").lower()` => "a vida é bela."               | Converte todos os caracteres da string em minúsculas.                                               |
| `str.upper()`              | `("A vida é bela.").upper()` => "A VIDA É BELA."               | Converte todos os caracteres da string em maiúsculas.                                               |
| `str.title()`              | `("A vida é bela.").title()` => "A Vida É Bela."               | Converte a primeira letra de cada palavra em maiúscula.                                              |
| `str.strip()`              | `(" A vida é bela. ").strip()` => "A vida é bela."              | Remove espaços em branco no início e no fim da string.                                               |
| `str.replace(a, b)`        | `("A vida é bela.").replace('vida', 'arte')` => "A arte é bela." | Substitui todas as ocorrências de `a` por `b` na string.                                              |
| `str.find(x)`              | `("A vida é bela.").find('vida')` => 2                         | Retorna o índice da primeira ocorrência de `x` na string, ou -1 se `x` não for encontrado.           |
| `str.count(x)`             | `("A vida é bela.").count('a')` => 3                           | Conta o número de ocorrências não sobrepostas de `x` na string.                                      |
| `str.startswith(x)`        | `("A vida é bela.").startswith('A')` => True                   | Retorna `True` se a string começar com o prefixo `x`.                                                |
| `str.endswith(x)`          | `("A vida é bela.").endswith('.')` => True                     | Retorna `True` se a string terminar com o sufixo `x`.                                                |
| `str.split(x)`             | `("A vida é bela.").split(' ')` => ['A', 'vida', 'é', 'bela.']  | Divide a string em uma lista onde cada palavra é uma string, usando o delimitador `x`.               |
| `str.join(x)`              | `(" ").join(['A', 'vida', 'é', 'bela.'])` => "A vida é bela."   | Concatena qualquer número de strings, com uma string `x` entre elas.                                 |


In [ ]:
s = "A vida é bela."

1. **`capitalize()`:** Usado para capitalizar a primeira letra da string.

In [ ]:
print(s.capitalize())

2. **`lower()`:** Usado para converter toda a string em letras minúsculas.

In [ ]:
print(s.lower())

3. **`upper()`:** Usado para converter toda a string em letras maiúsculas.

In [ ]:
print(s.upper())

4. **`title()`:** Usado para converter a primeira letra de cada palavra em maiúscula.

In [ ]:
print(s.title())

5. **`strip()`:** Usado para remover espaços em branco no início e no fim da string.

In [ ]:
s_com_espacos = "  A vida é bela.  "
print(s_com_espacos.strip())

A vida é bela.


6. **`replace(a, b)`:** Usado para substituir todas as ocorrências de `a` por `b`.

In [ ]:
print(s.replace('vida', 'arte'))

7. **`find(x)`:** Usado para encontrar a primeira ocorrência de `x`.

In [ ]:
print(s.find('vida'))

8. **`count(x)`:** Usado para contar o número de ocorrências de `x`.

In [ ]:
print(s.count('a'))

9. **`startswith(x)`:** Usado para verificar se a string começa com `x`.

In [ ]:
print(s.startswith('A'))

10. **`endswith(x)`:** Usado para verificar se a string termina com `x`.

In [ ]:
print(s.endswith('.'))

11. **`split(x)`:** Usado para dividir a string em partes usando `x` como delimitador.

In [ ]:
print(s.split(' '))

12. **`join(x)`:** Usado para juntar uma lista de strings usando `x` como conector.

In [ ]:
words = ['A', 'vida', 'é', 'bela.']
print(" ".join(words))  # Saída: "A vida é bela."

In [ ]:
# Usando Métodos de Strings
texto = '   Python é Poderoso!   '

# Convertendo para Maiúsculas e Minúsculas
maiusculas = texto.upper()
minusculas = texto.lower()

# Removendo Espaços em Branco
sem_espacos = texto.strip()

# Substituindo Texto
substituido = texto.replace('Poderoso', 'Incrível')

# Encontrando Substring
posicao = texto.find('é')

# Dividindo e Unindo Strings
dividido = texto.split()
unido = ' '.join(dividido)

maiusculas, minusculas, sem_espacos, substituido, posicao, dividido, unido

## Formatação de Strings
A formatação de strings permite inserir valores dentro de uma string de maneira dinâmica. Em Python, podemos usar f-strings ou o método `format()`.

In [ ]:
# Usando f-strings
nome = 'Maria'
idade = 30
f_string = f'{nome} tem {idade} anos.'

# Usando o método format()
format_string = '{} tem {} anos.'.format(nome, idade)

f_string, format_string

#### Formato de String: permite a inserção e formatação de variáveis dentro de uma string, útil para relatórios e visualizações.

##### Formatação de Ponto Flutuante

In [17]:
"O preço é {:.2f}".format(5.678)

'O preço é 5.68'

O código `"O preço é {:.2f}".format(5.678)` está usando a formatação de strings em Python para formatar o valor de ponto flutuante `5.678` com duas casas decimais.

1. **String Literal**:
   ```python
   "O preço é {:.2f}"
   ```
   - `"O preço é "`: Esta é a parte fixa da string que será impressa como está.
   - `{}`: As chaves são usadas para inserir um valor na string. O valor a ser inserido é fornecido pelo método `format`.
   - `:`: Inicia a especificação de formato dentro das chaves.
   - `.2`: Define o número de casas decimais como 2.
   - `f`: Especifica que o valor deve ser formatado como um número de ponto flutuante.

2. **Método `format`**:
   ```python
   .format(5.678)
   ```
   O método `format` é chamado na string, e o valor `5.678` é passado como argumento. O valor é então formatado de acordo com a especificação dentro das chaves `{}`.

3. **Resultado**:
   O resultado é a seguinte string:
   ```
   'O preço é 5.68'
   ```
   O valor `5.678` foi arredondado para duas casas decimais, resultando em `5.68`.



##### Formatação de Números Inteiros:
Para imprimir um número de ordem com um número fixo de dígitos, preenchendo com zeros:

In [26]:
ordem = 42
print("Ordem do pedido: {:05d}".format(ordem))

Ordem do pedido: 00042


1. **Atribuição de Valor**:
   ```python
   ordem = 42
   ```
   A variável `ordem` é atribuída o valor inteiro `42`, que representa o número de uma ordem específica em um contexto, como um pedido em um restaurante ou uma compra online.

2. **String Literal com Formatação**:
   ```python
   "Ordem do pedido: {:05d}"
   ```
   - `"Ordem do pedido: "`: Essa é a parte fixa da string que será impressa como está.
   - `{}`: As chaves são usadas para inserir um valor na string. O valor a ser inserido é fornecido pelo método `format`.
   - `:`: Inicia a especificação de formato dentro das chaves.
   - `0`: Indica que o preenchimento deve ser feito com zeros, se necessário.
   - `5`: Define a largura mínima do campo como 5 caracteres.
   - `d`: Especifica que o valor é um número inteiro decimal (dígito).

3. **Método `format`**:
   ```python
   .format(ordem)
   ```
   O método `format` é chamado na string, e o valor da variável `ordem` é passado como argumento. O valor é então formatado de acordo com a especificação dentro das chaves `{}`.

4. **Resultado**:
   O resultado é a seguinte string:
   ```
   'Ordem do pedido: 00042'
   ```
   Como o especificador de formato é `05d`, o número `42` é formatado com cinco dígitos, preenchendo com zeros à esquerda para atingir esse tamanho.



##### Formatação de Porcentagem

In [31]:
progresso = 0.456
print("Progresso do projeto: {:.3%}".format(progresso))


Progresso do projeto: 45.600%


**Detalhamento do método:**

1. **Atribuição de Valor**:
   ```python
   progresso = 0.456
   ```
   A variável `progresso` recebe o valor `0.456`, que representa 45.6% em termos de porcentagem.

2. **Formatação da String**:
   ```python
   "Progresso do projeto: {:.3%}".format(progresso)
   ```
   - `"Progresso do projeto: "`: Esta é a parte fixa da string que será impressa como está.
   - `{}`: As chaves são usadas para inserir um valor na string. O valor a ser inserido é fornecido pelo método `format`.
   - `:`: Inicia a especificação de formato.
   - `.3`: Define o número de casas decimais para 3.
   - `%`: Especifica que o valor deve ser formatado como uma porcentagem. Isso multiplica o valor por 100 e adiciona o símbolo de porcentagem (`%`) após o número.

3. **Método `format`**:
   O método `format` é chamado na string, e o valor de `progresso` é passado como argumento. O valor é então formatado de acordo com a especificação dentro das chaves `{}`.

4. **Resultado**:
   O resultado é a seguinte string:
   ```
   'Progresso do projeto: 45.600%'
   ```
   Essa string é impressa no console pelo comando `print`.


#####Combinação de Argumentos:


In [37]:
print("{0} tem {1:.2f} anos e {2:.2%} de chance de ser aprovada para a vaga".format("Alice", 25, 0.75))


Alice tem 25.00 anos e 75.00% de chance de ser aprovada para a vaga


##### Formatação de Dinheiro

In [40]:
preco = 12345.678
print("Preço: ${:,.2f}".format(preco))

Preço: $12,345.68


E o padrão brasileiro?

In [45]:
preco = 12345.678

# Utiliza o método format para arredondar para duas casas decimais
preco_formatado = "{:,.2f}".format(preco)

# Substitui o ponto por vírgula para o separador decimal e a vírgula por ponto para o separador de milhares
preco_formatado = preco_formatado.replace(',', ' ').replace('.', ',').replace(' ', '.')

print("Preço: R${}".format(preco_formatado))

Preço: R$12.345,68


## Manipulação de Strings com combinação de métodos

### `capitalize()` e `title()`: Capitalizar cada frase em um parágrafo.

In [ ]:
paragrafo = "a vida é bela. a arte é admirável. aprendemos todos os dias."

In [ ]:
sentences = paragrafo.split('. ')

In [ ]:
capitalized_sentences = [sentence.capitalize() for sentence in sentences]

In [ ]:
novo_paragrafo = ". ".join(capitalized_sentences)

In [ ]:
print(novo_paragrafo)

A vida é bela. A arte é admirável. Aprendemos todos os dias.


### `lower()` e `count()`: Contar o número de ocorrências de uma palavra, independentemente do caso.

In [ ]:
word_to_count = "vida"

In [ ]:
count = s.lower().count(word_to_count.lower())

In [ ]:
print(f"A palavra '{word_to_count}' aparece {count} vezes.")


A palavra 'vida' aparece 1 vezes.


### `upper()` e `startswith()`: Transformar em maiúsculas as palavras que começam com uma letra específica.

In [ ]:
words = s.split(' ')
words

['A', 'vida', 'é', 'bela.']

In [ ]:
for word in words:
    if word.startswith('v'):
        print(word.upper())

VIDA


### `replace(a, b)` com condicional: Substituir uma palavra somente se ela existir na string.

In [ ]:
old_word = "vida"
new_word = "arte"

In [ ]:
if old_word in s:
    print(s.replace(old_word, new_word))

A arte é bela.


### `split(x)` e `join(x)` com loop: Dividir uma string em palavras e reverter a ordem das palavras.

In [ ]:
words = s.split(' ')
words

['A', 'vida', 'é', 'bela.']

A função `reversed()` em Python é usada para inverter a ordem dos elementos de um objeto iterável, como uma lista ou string. Ela não modifica o objeto original, mas retorna um iterador que percorre os elementos na ordem invertida.

In [ ]:
reversed_words = [word for word in reversed(words)]
reversed_words

['bela.', 'é', 'vida', 'A']

In [ ]:
reversed_sentence = " ".join(reversed_words)

In [ ]:
print(reversed_sentence)

bela. é vida A


In [ ]:
strings = [" a Vida é bela ", " a Arte é admirável ", " aprendemos todos os dias "]
for string in strings:
    cleaned_string = string.strip().lower().replace('a', '@')
    print(cleaned_string)


### Pre-processamento de texto

#### Extraindo o usuário  e o servidor do e_mail

In [1]:
emails = [
    "alice@example.com",
    "bob@provider.com",
    "carol@domain.net",
    "david@service.org"
]

In [3]:
usuarios_email = [email.split('@')[0] for email in emails]
print(usuarios_email)

['alice', 'bob', 'carol', 'david']


In [5]:
servidores_email = [email.split('@')[1] for email in emails]
print(servidores_email)

['example.com', 'provider.com', 'domain.net', 'service.org']


In [9]:
[servidores.split('.')[0] for servidores in servidores_email]

['example', 'provider', 'domain', 'service']

In [15]:
palavra = 'saudade'

In [16]:
list(palavra)

['s', 'a', 'u', 'd', 'a', 'd', 'e']

#### O pré-processamento de texto é uma etapa essencial em muitas tarefas de Processamento de Linguagem Natural (NLP). Ele envolve a limpeza e transformação do texto bruto em uma forma que pode ser analisada e modelada de maneira mais eficaz. Abaixo são apresentados exemplos comuns de pré-processamento em NLP: